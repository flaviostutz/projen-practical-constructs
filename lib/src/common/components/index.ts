export * from './base-tooling';
export * from './common-target-type';
export * from './common-targets';
export * from './release-tasks';
export * from './publish';
