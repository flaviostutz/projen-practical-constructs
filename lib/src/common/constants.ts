export const PROJEN_VERSION = '0.91.6';
export const NODE_VERSION = '20.16.0';
