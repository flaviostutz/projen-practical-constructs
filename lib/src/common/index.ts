export * from './components';
export * from './files';
export * from './types';
