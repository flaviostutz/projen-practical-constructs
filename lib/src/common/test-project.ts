import { mkdtempSync } from 'node:fs';
import { tmpdir } from 'node:os';
// eslint-disable-next-line unicorn/import-style
import { join } from 'node:path';

import { Project } from 'projen';

export class TestProject extends Project {
  constructor() {
    const outdir = mkdtempSync(join(tmpdir(), 'test-'));
    super({ name: 'test', outdir });
  }
}
