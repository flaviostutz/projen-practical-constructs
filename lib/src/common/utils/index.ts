export * from './ts2jsii';
export * from './monotag';
