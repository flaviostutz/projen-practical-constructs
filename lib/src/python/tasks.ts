import { Project, Task } from 'projen';

export interface TaskOptionsTarget {
  /**
   * Path to the python virtual environment directory
   * used in this project
   * @default .venv
   */
  readonly venvPath?: string;
  /**
   * Existing task to attach new tasks to.
   * It will be included as "spawn" tasks in new steps
   */
  readonly attachTasksTo?: string;
}

export interface TaskOptions {
  /**
   * Path to the python virtual environment directory
   * used in this project
   */
  readonly venvPath: string;
}

export const addSpawnTaskToExisting = (
  project: Project,
  task: Task,
  parentTaskName?: string,
): void => {
  if (!parentTaskName) {
    return;
  }
  const attachTask = project.tasks.tryFind(parentTaskName);
  if (!attachTask) {
    throw new Error(
      `Spawn task '${task.name}' cannot be configured: Parent task '${parentTaskName}' doesn't exist in project globals`,
    );
  }
  attachTask.spawn(task);
};
