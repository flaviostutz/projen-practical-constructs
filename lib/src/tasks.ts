import { Project, Task } from 'projen';

export interface TaskOptions {
  /**
   * Path to the python virtual environment directory
   * used in this project
   * @default .venv
   */
  readonly venvPath?: string;
  /**
   * Existing task to attach new tasks to.
   * It will be included as "spawn" tasks in new steps
   */
  readonly attachTasksTo?: string;
}

export const addTaskToParent = (project: Project, task: Task, parentTaskName?: string): void => {
  if (!parentTaskName) {
    return;
  }
  const attachTask = project.tasks.tryFind(parentTaskName);
  if (!attachTask) {
    throw new Error(`'${parentTaskName}' task not found`);
  }
  attachTask.spawn(task);
};
