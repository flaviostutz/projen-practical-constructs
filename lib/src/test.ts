/* eslint-disable no-restricted-syntax */
/* eslint-disable no-new */
import { Component, Project } from 'projen';

import { PyTest, PyTestOptions } from './components/pytest';
import { TaskOptions } from './tasks';

/**
 * Python project lint configurations
 */
export class TestTarget extends Component {
  constructor(project: Project, taskOpts: TaskOptions, opts?: TestOptions) {
    super(project);

    project.tasks.addTask('test', {
      description: `Test project (unit tests/coverage)`,
    });

    new PyTest(project, taskOpts, opts);
  }
}

export interface TestOptions extends PyTestOptions {}
