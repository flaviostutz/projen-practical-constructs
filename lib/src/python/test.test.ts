/* eslint-disable no-new */
import { Testing } from 'projen';

import { TestProject } from '../common/test-project';

import { TestTarget } from './test';
import { TaskOptions } from './tasks';

test('TestTarget is synthesized correctly', () => {
  const project = new TestProject();
  const taskOpts: TaskOptions = { venvPath: '.venv' };
  project.removeTask('build');
  project.removeTask('test');
  new TestTarget(project, taskOpts);
  const output = Testing.synth(project);
  expect(output).toMatchSnapshot();
});
