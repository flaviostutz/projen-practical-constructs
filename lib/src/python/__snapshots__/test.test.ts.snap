// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TestTarget is synthesized correctly 1`] = `
{
  ".coveragerc": "[run]
branch = true
omit = 
  

[report]
fail_under = 80
format = text
skip_empty = true
skip_covered = false",
  ".gitattributes": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

* text=auto eol=lf",
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
node_modules/
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
!/pytest.ini
!/.coveragerc
.coverage
.pytest_cache
",
  ".projen/deps.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "dependencies": [
      {
        "name": "pytest-cov",
        "type": "devenv",
        "version": "6.0.*",
      },
      {
        "name": "pytest",
        "type": "devenv",
        "version": "8.3.*",
      },
    ],
  },
  ".projen/files.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "files": [
      ".coveragerc",
      ".gitattributes",
      ".gitignore",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
      "pytest.ini",
    ],
  },
  ".projen/tasks.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "tasks": {
      "compile": {
        "description": "Only compile",
        "name": "compile",
      },
      "default": {
        "description": "Synthesize project files",
        "name": "default",
      },
      "eject": {
        "description": "Remove projen from the project",
        "env": {
          "PROJEN_EJECTING": "true",
        },
        "name": "eject",
        "steps": [
          {
            "spawn": "default",
          },
        ],
      },
      "package": {
        "description": "Creates the distribution package",
        "name": "package",
      },
      "post-compile": {
        "description": "Runs after successful compilation",
        "name": "post-compile",
      },
      "pre-compile": {
        "description": "Prepare the project for compilation",
        "name": "pre-compile",
      },
      "test": {
        "description": "Test project (unit tests/coverage)",
        "name": "test",
      },
      "test-unit": {
        "description": "Unit tests (pytest)",
        "name": "test-unit",
        "steps": [
          {
            "exec": ".venv/bin/pytest",
          },
        ],
      },
    },
  },
  "pytest.ini": "[pytest]
cache_dir = .cache/.pytest_cache
addopts = --cov=src --verbose
testpaths =
    tests
    integration
",
}
`;
